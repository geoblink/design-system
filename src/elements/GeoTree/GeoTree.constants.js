const INPUT_MODES = {
  SINGLE: 'single',
  MULTIPLE: 'multiple',
  HIDDEN: 'hidden'
}

export {
  INPUT_MODES
}
