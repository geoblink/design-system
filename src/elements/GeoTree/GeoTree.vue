<template>
  <div class="geo-tree">
    <geo-bordered-box-header-search-form
      v-if="searchable"
      v-model="searchQuery"
      :placeholder="searchPlaceholder"
      @input="handleSearching"
    />
    <geo-scrollable-container>
      <div
        v-if="isLoading"
        class="geo-tree__loading"
        v-text="loadingLabel"
      />
      <div
        v-if="noResultsFound"
        class="geo-tree__no-results-found"
        v-text="noResultsFoundLabel"
      />
      <ul
        v-else
        class="geo-tree__list"
      >
        <geo-tree-item
          v-for="category in sortedCategories"
          :key="category[keyForId]"
          :category="category"
          :key-for-id="keyForId"
          :key-for-label="keyForLabel"
          :checked-items="checkedItems"
          :key-for-children="keyForChildren"
          :description-icon="descriptionIcon"
          @click="handleCategoryClick"
          @check="handleCheckItem"
        >
          <template v-slot:trailingAccessoryAction>
            <slot name="actionButton" />
          </template>
        </geo-tree-item>
      </ul>
    </geo-scrollable-container>
  </div>
</template>

<script>
import _ from 'lodash'
import fuzzAldrin from 'fuzzaldrin-plus'

export default {
  name: 'GeoTree',
  status: 'ready',
  release: '29.9.0',
  props: {
    /**
     * Text to display as placeholder of the search input
   */
    searchPlaceholder: {
      type: String,
      required: false
    },
    /**
       * Text to display when it's loading data
   */
    loadingLabel: {
      type: String,
      require: false
    },
    /**
       * Text to display when no results found on searching
   */
    noResultsFoundLabel: {
      type: String,
      require: false
    },
    /**
     * Categories to be displayed in the tree component
     * Structure:
     * [
     *     id,
     *     label,
     *     children: [
     *      id,
     *      label,
     *      children
     *     ]
     * ]
     */
    categories: {
      type: Array,
      required: true
    },
    /**
     * Key to access to the identifier of the item
     */
    keyForId: {
      type: String,
      required: false,
      default: 'id'
    },
    /**
     * Key to access to the label of the item
     */
    keyForLabel: {
      type: String,
      required: false,
      default: 'label'
    },
    /**
     * Key to access to the children items of the category
     */
    keyForChildren: {
      type: String,
      required: false,
      default: 'children'
    },
    /**
    * Used determine if you can search or not
    */
    searchable: {
      type: Boolean,
      required: false,
      default: false
    },
    /**
     * Initial selected items
    */
    initialState: {
      type: Object,
      required: false
    },
    /**
     * Icon used to alert about some extra info displayed in a popover
     */
    descriptionIcon: {
      type: Array,
      required: false
    },
    /**
     * Should appear all items collapsed or not initially
     */
    isInitiallyExpanded: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  data () {
    return {
      searchQuery: '',
      isLoading: false,
      checkedItems: {},
      filteredCategories: []
    }
  },
  computed: {
    sortedCategories () {
      return _.sortBy(this.filteredCategories, [this.keyForLabel])
    },
    noResultsFound () {
      return this.searchQuery && !this.filteredCategories.length
    }
  },
  mounted () {
    this.setInitialState(this.initialState)

    if (this.categories) {
      const setCategoriesNoExpanded = (category) => {
        _.forEach(category[this.keyForChildren], innerCategory => {
          setCategoriesNoExpanded(innerCategory)
        })

        return _.assign({}, category, { isExpanded: this.isInitiallyExpanded })
      }

      this.filteredCategories = _.map(this.categories, setCategoriesNoExpanded)
    }
  },
  methods: {
    handleSearching () {
      this.$emit('search', this.searchQuery)

      const clearString = string => _.toLower(_.deburr(string))

      const getFilteredCategories = (categories, query) => categories.reduce((carry, category) => {
        const isCategoryMatching = fuzzAldrin.score(category[this.keyForLabel], query) > 0
        const categories = category[this.keyForChildren] ? getFilteredCategories(category[this.keyForChildren], query) : null

        if (categories && categories.length) {
          return [
            ...carry,
            _.assign(
              {},
              category,
              {
                [this.keyForChildren]: categories,
                isExpanded: true,
                matches: fuzzAldrin.match(clearString(category[this.keyForLabel]), clearString(query))
              }
            )
          ]
        }
        return isCategoryMatching
          ? [
            ...carry,
            _.assign(category, { matches: fuzzAldrin.match(clearString(category[this.keyForLabel]), clearString(query)) })
          ]
          : carry
      }, [])

      this.filteredCategories = this.searchQuery
        ? getFilteredCategories(this.categories, this.searchQuery)
        : this.categories
    },
    handleCategoryClick (clickedCategory) {
      const isExpanded = (category) => category[this.keyForId] === clickedCategory[this.keyForId] ? !category.isExpanded : category.isExpanded
      const toggleCategoriesExpanded = (categories) => _.map(categories, category => {
        const children = category[this.keyForChildren] ? toggleCategoriesExpanded(category[this.keyForChildren]) : null

        return _.assign(
          {},
          category,
          {
            isExpanded: isExpanded(category),
            [this.keyForChildren]: children
          }
        )
      })
      this.filteredCategories = toggleCategoriesExpanded(this.filteredCategories)

      this.$emit('click', clickedCategory)
    },
    handleCheckItem (category, isChecked) {
      const categoryId = category[this.keyForId]
      this.checkedItems = isChecked
        ? _.assign({}, this.checkedItems, { [categoryId]: true })
        : _.omit(this.checkedItems, categoryId)
    },
    setInitialState (initialState) {
      if (!initialState) return

      _.assign(this.checkedItems, initialState)
    }
  }
}
</script>
