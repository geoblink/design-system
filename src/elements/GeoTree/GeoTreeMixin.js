/**
 * @mixin
 */

const _ = require('lodash')
export default {
  props: {
    /**
     * Object needed in case you want to support drag elements
     *
     * You have to use the same format as group prop of [vuedraggable](https://www.npmjs.com/package/vuedraggable)
     * for more info about this.
     */
    draggableGroup: {
      type: Object,
      required: false
    }
  },
  computed: {
    // Vuedraggable library allows to visually pull elements even with pull disabled, we don't want that
    dragClassToIgnore () {
      if (!this.group) return ''

      return !this.group.pull
        ? 'geo-tree__drag-disabled'
        : ''
    },

    hasMoreResultsToLoad () {
      return this.pageSize * this.visiblePages < this.filteredCategories.length
    },

    visibleItems () {
      if (this.hasMoreResultsToLoad) return _.slice(this.filteredCategories, 0, this.visiblePages * this.pageSize)
      return this.filteredCategories
    }
  },
  methods: {
    startDrag ($event) {
      this.$emit('start-drag', $event)
    },
    endDrag ($event) {
      this.$emit('end-drag', $event)
    },
    emitChangeDrag ($event) {
      this.$emit('change-drag', $event)
    },
    changeDrag ($event, parentCategory) {
      this.emitChangeDrag({
        event: $event,
        parentCategory
      })
    },
    loadNextPage (payload, pageObject = 'visiblePages') {
      _.set(this, pageObject, _.get(this, pageObject) + 1)
      this.$nextTick(function () {
        payload.scrollToLastEntry()
      })
    }
  }
}
