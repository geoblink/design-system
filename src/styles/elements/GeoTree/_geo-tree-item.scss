.geo-tree-item {
  cursor: pointer;
}

.geo-tree-item label {
  cursor: pointer;
}

.geo-tree-item__list {
  list-style: none;
}

.geo-tree-item .geo-list-item__icon-and-label__icon-container {
  text-align: center;
  width: 14px;
}

.geo-tree-item .geo-list-item__icon-and-label__icon-container__icon {
  color: $color-grey-backdrop;
  width: 5px;
}

.geo-tree-item__total-items {
  @include font($body-small);
  color: $color-grey-light;
  font-style: italic;
}

.geo-tree-item .geo-list-item__icon-and-label__icon-container__icon {
  transition: transform 0.25s;
}

.geo-tree-item--expanded .geo-list-item__icon-and-label__icon-container__icon {
  transform: rotate(90deg);
}

.geo-tree-item--single {
  margin-left: $space-small;
}
