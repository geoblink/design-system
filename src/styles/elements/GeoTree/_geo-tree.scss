.geo-tree__list {
  list-style: none;
}
