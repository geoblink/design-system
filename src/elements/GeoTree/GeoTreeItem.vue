<template>
  <li
    class="geo-tree-item"
  >
    <geo-list-item
      :class="{
        'geo-tree-item--rotated-icon': hasToRotateIcon,
        'geo-tree-item--single': isSingleItem
      }"
      :icon="categoryIcon"
      @click="handleClick"
    >
      <label class="geo-tree-item__label">
        <geo-highlighted-string
          :highlighted-chars="category.matches"
          :reference-string="category[keyForLabel]"
        />
        <span
          v-if="!isSingleItem"
          class="geo-tree-item__total-items"
        >
          ({{ totalCategoryChildren }})
        </span>
        <span
          v-if="category.description"
          class="geo-tree-item__description"
        >
          <font-awesome-icon
            :icon="descriptionIcon"
            fixed-width
          />
          <geo-tooltip>
            {{ category.description }}
          </geo-tooltip>
        </span>
      </label>
      <template slot="trailingAccessoryItem">
        <span>
          <slot
            name="trailingAccessoryAction"
            :item="category"
          />
        </span>
        <input
          :id="category[keyForId]"
          :checked="isChecked"
          :indeterminate.prop="isIndeterminate"
          :disabled="isInputDisabled"
          type="checkbox"
          @click.stop
          @input="handleCheck(category, $event.target.checked)"
        >
      </template>
    </geo-list-item>
    <ul
      v-if="isExpanded"
      class="geo-tree-item__list"
    >
      <draggable
        :list="category[keyForSubcategory]"
        :group="draggableGroup"
        :sort="false"
        :disabled="!draggableGroup"
        :filter="`.${dragClassToIgnore}`"
        drag-class="geo-tree__dragger"
        ghost-class="geo-tree__ghost"
        @start="startDrag($event)"
        @end="endDrag($event)"
        @change="changeDrag($event, category)"
      >
        <geo-tree-item
          v-for="subcategory in category[keyForSubcategory]"
          :key="subcategory[keyForId]"
          :class="dragClassToIgnore"
          :data-test="`subcategory-${subcategory[keyForId]}`"
          :category="subcategory"
          :key-for-id="keyForId"
          :key-for-label="keyForLabel"
          :key-for-subcategory="keyForSubcategory"
          :expanded-categories="expandedCategories"
          :checked-items="checkedItems"
          :collapsed-icon="collapsedIcon"
          :expanded-icon="expandedIcon"
          :description-icon="descriptionIcon"
          :draggable-group="draggableGroup"
          @check-item="handleCheckChildItem"
          @check-folder="handleCheckChildFolder"
          @click="handleClick"
          @toggleExpand="toggleExpand"
          @start-drag="startDrag($event)"
          @end-drag="endDrag($event)"
          @change-drag="emitChangeDrag($event)"
        >
          <template
            slot="trailingAccessoryAction"
            slot-scope="{ item }"
          >
            <slot
              name="trailingAccessoryAction"
              :item="item"
            />
          </template>
        </geo-tree-item>
      </draggable>
    </ul>
  </li>
</template>

<script>
import _ from 'lodash'
import Draggable from 'vuedraggable'

import GeoTreeMixin from './GeoTreeMixin'

export default {
  name: 'GeoTreeItem',
  status: 'ready',
  release: '29.9.0',
  components: {
    Draggable
  },
  mixins: [GeoTreeMixin],
  props: {
    category: {
      type: Object,
      required: true
    },
    /**
     * Key to access to the identifier of the item
     */
    keyForId: {
      type: String,
      required: false,
      default: 'id'
    },
    /**
     * Key to access to the label of the item
     */
    keyForLabel: {
      type: String,
      required: false,
      default: 'label'
    },
    /**
     * Key to access to the children items of the category
     */
    keyForSubcategory: {
      type: String,
      required: false,
      default: 'subcategory'
    },
    /**
     * List of all the items checked, it's an object with truthy keys
     */
    checkedItems: {
      type: Object,
      required: true
    },
    /**
     * Icon used to alert about some extra info displayed in a popover
     */
    descriptionIcon: {
      type: Array,
      default: () => []
    },
    /*
    * List of categories that should appear expanded
    * */
    expandedCategories: {
      type: Object,
      required: true
    },
    /**
     * Optional Font Awesome 5 icon to use as collapsed icon
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    collapsedIcon: {
      type: Array,
      required: true,
      default: function () {
        return ['fal', 'chevron-right']
      }
    },
    /**
     * Optional Font Awesome 5 icon to use as expanded icon
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    expandedIcon: {
      type: Array,
      required: false
    }
  },
  data () {
    return {
      isShowActionButton: false
    }
  },
  computed: {
    categoryIcon () {
      if (this.isSingleItem) return null

      return this.isExpanded
        ? this.expandedIcon || this.collapsedIcon
        : this.collapsedIcon
    },
    isIndeterminate () {
      const isSomeChildSelected = category => {
        return _.some(category[this.keyForSubcategory], subCategory => {
          if (subCategory[this.keyForSubcategory]) return isSomeChildSelected(subCategory)

          return !!this.checkedItems[subCategory[this.keyForId]]
        })
      }

      return !this.isChecked && isSomeChildSelected(this.category)
    },
    isChecked () {
      if (this.isInputDisabled) return false

      const allAreChildrenSelected = category => {
        const selectableChildren = this.getSelectableChildrenForCategory(category)
        return _.every(selectableChildren, subCategory => {
          if (_.size(subCategory[this.keyForSubcategory])) return allAreChildrenSelected(subCategory)

          return !!this.checkedItems[subCategory[this.keyForId]]
        })
      }
      return !this.isSingleItem
        ? allAreChildrenSelected(this.category)
        : !!this.checkedItems[this.category[this.keyForId]]
    },
    totalCategoryChildren () {
      const sumOfChildren = category => _.size(category[this.keyForSubcategory]) + _.sumBy(category[this.keyForSubcategory], sumOfChildren)

      return sumOfChildren(this.category)
    },
    totalSelectableCategoryChildren () {
      const sumOfSelectableChildren = category => {
        const selectableChildren = this.getSelectableChildrenForCategory(category)
        return _.size(selectableChildren) + _.sumBy(category[this.keyForSubcategory], sumOfSelectableChildren)
      }

      return sumOfSelectableChildren(this.category)
    },
    isExpanded () {
      return !!this.expandedCategories[this.category[this.keyForId]]
    },
    isSingleItem () {
      return _.isNil(this.category[this.keyForSubcategory])
    },
    hasToRotateIcon () {
      return this.isExpanded && !this.expandedIcon
    },
    isInputDisabled () {
      return !this.isSingleItem && this.totalSelectableCategoryChildren === 0
    }
  },
  methods: {
    /**
     * On list item click
     */
    handleClick () {
      if (this.isSingleItem) {
        this.handleCheck(this.category, !this.isChecked)
      } else {
        this.toggleExpand(this.category)
      }
    },
    handleCheckChildItem (category, isChecked, isDelegated) {
      this.$emit('check-item', category, isChecked, isDelegated)
    },
    handleCheckChildFolder (category, isChecked, isDelegated) {
      this.$emit('check-folder', category, isChecked, isDelegated)
    },
    /**
     * To check all items of a category
     */
    handleCheck (category, isChecked, isDelegated = false) {
      if (category[this.keyForSubcategory]) {
        if (_.size(category[this.keyForSubcategory])) { // We don't want to emit event for empty folders
          this.$emit('check-folder', _.omit(category, this.keyForSubcategory), isChecked, isDelegated)
        }
        _.forEach(category[this.keyForSubcategory], (innerCategory) => {
          this.handleCheck(innerCategory, isChecked, true)
        })
      } else {
        this.$emit('check-item', category, isChecked, isDelegated)
      }
    },
    toggleExpand (category) {
      this.$emit('toggleExpand', category)
    },
    getSelectableChildrenForCategory (category) {
      const keyForSubcategory = this.keyForSubcategory
      return getSelectableChildrenRecursively(category)

      function getSelectableChildrenRecursively (currentCategory) {
        return _.reject(currentCategory[keyForSubcategory], subCategory => {
          return subCategory[keyForSubcategory] &&
            (!_.size(subCategory[keyForSubcategory]) || !_.size(getSelectableChildrenRecursively(subCategory)))
        })
      }
    }
  }
}
</script>
