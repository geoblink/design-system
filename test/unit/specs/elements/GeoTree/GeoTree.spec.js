import _ from 'lodash'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import GeoTree from '@/elements/GeoTree/GeoTree.vue'

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.component('geo-tree', GeoTree)

const CATEGORIES = [
  {
    id: 'fruits',
    label: 'Fruits',
    subcategories: [
      {
        id: 'tropical-fruits',
        label: 'Tropical fruits',
        subcategories: [
          {
            id: 'pineapple',
            label: 'Pineapple'
          },
          {
            id: 'banana',
            label: 'Banana'
          },
          {
            id: 'coconut',
            label: 'Coconut'
          },
          {
            id: 'avocado',
            label: 'Avocado'
          }
        ]
      },
      {
        id: 'citrus-fruits',
        label: 'Citrus fruits',
        subcategories: [
          {
            id: 'orange',
            label: 'Orange'
          },
          {
            id: 'lime',
            label: 'Lime'
          },
          {
            id: 'grapefruit',
            label: 'GrapeFruit'
          },
          {
            id: 'mandarin',
            label: 'Mandarin'
          },
          {
            id: 'pomelo',
            label: 'Pomelo'
          }
        ]
      },
      {
        id: 'sweet-fruits',
        label: 'Sweet fruits',
        subcategories: [
          {
            id: 'pear',
            label: 'Pear'
          },
          {
            id: 'apple',
            label: 'Apple'
          },
          {
            id: 'redGrapes',
            label: 'Red Grapes'
          }
        ]
      }
    ]
  },
  {
    id: 'vegetables',
    label: 'Vegetables',
    subcategories: [
      {
        id: 'fruits',
        label: 'Fruits',
        subcategories: [
          {
            id: 'eggplant',
            label: 'Eggplant'
          },
          {
            id: 'pepper',
            label: 'Pepper'
          }
        ]
      },
      {
        id: 'bulbs',
        label: 'Bulbs',
        subcategories: [
          {
            id: 'onion',
            label: 'Onion'
          },
          {
            id: 'leek',
            label: 'Leek'
          },
          {
            id: 'garlic',
            label: 'Garlic'
          }
        ]
      }
    ]
  }
]

describe.only('GeoTree', () => {
  const getWrapper = props => shallowMount(GeoTree, {
    propsData: _.assign(
      {},
      {
        keyForLabel: 'label',
        keyForChildren: 'subcategories',
        keyForId: 'id',
        categories: CATEGORIES
      },
      props
    )
  })

  it('should render correctly', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  it('should render a list of categories which length is equal to the categories prop', () => {
    expect(getWrapper().findAll('geo-tree-item').length).toEqual(CATEGORIES.length)
  })

  it('should display a loading label when isLoading prop is passed', () => {
    const loadingLabel = 'Fake loading label'
    const wrapper = getWrapper({
      isLoading: true,
      loadingLabel
    })

    expect(wrapper.find('.geo-tree__loading').exists()).toBeTruthy()
    expect(wrapper.find('.geo-tree__loading').text()).toEqual(loadingLabel)
  })

  it('should render a searcher box when searchable prop is passed', () => {
    const wrapper = getWrapper({
      searchable: true
    })

    expect(wrapper.find('geo-bordered-box-header-search-form').exists()).toBeTruthy()
  })

  it('should filter the categories when on user searching', () => {
    const wrapper = getWrapper({
      searchable: true
    })
    const expectedFilteredCategories = []

    /*
    * TODO:
    *  1. Search for the searcher input.
    *  2. Type some text that can be found in the categories.
    *
    * Â¿? How to know that filtered categories are right?
    */

    expect(getWrapper().findAll('geo-tree-item').length).toEqual(expectedFilteredCategories.length)
  })

  it('should display a no results found label when nothing matches with searched text', () => {
    const noResultsLabel = 'Fake no results label'
    const wrapper = getWrapper({
      searchable: true,
      noResultsLabel
    })

    /*
    * TODO:
    *  1. Search for the searcher input.
    *  2. Type some random weird text
    * */

    expect(wrapper.find('.geo-tree__no-results-found').exists()).toBeTruthy()
    expect(wrapper.find('.geo-tree__no-results-found').text()).toEqual(noResultsLabel)
  })
})
