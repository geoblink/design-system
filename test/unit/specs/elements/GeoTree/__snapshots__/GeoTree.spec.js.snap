// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoTree should render correctly 1`] = `
<div class="geo-tree">
  <!---->
  <geo-scrollable-container>
    <!---->
    <ul class="geo-tree__list"></ul>
  </geo-scrollable-container>
</div>
`;
